misc/release.py

```diff
@@ -97,14 +100,53 @@ def link(sourcepath, destpath):
         os.remove(destpath)
     os.link(sourcepath, destpath)
 
-def compile(goos=None, goarch=None):
+_tagpat = re.compile(r'tag:\s+([^,]+)')
+
+def compile_version_opts(release_version=None, allow_mismatch=False):
+    result = subprocess.run(['git', 'log', '-n', '1', '--pretty=%H %D'], capture_output=True)
+    result.check_returncode()
+    so = result.stdout.decode()
+    githash, desc = so.split(None, 1)
+    tags = []
+    tag = None
+    for m in _tagpat.finditer(desc):
+        tag = m.group(1)
+        tags.append(tag)
+        if tag == release_version:
+            break
+    if tag != release_version:
+        if not allow_mismatch:
+            raise Exception('.version is {!r} but tags {!r}'.format(release_version, tags))
+        else:
+            logger.warning('.version is %r but tags %r', release_version, tags)
+    now = time.strftime('%Y-%m-%dT%H:%M:%S', time.gmtime()) + '+0000'
+    result = subprocess.run(['git', 'status', '--porcelain'], capture_output=True)
+    result.check_returncode()
+    if len(result.stdout) > 2:
+        dirty = "true"
+    else:
+        dirty = ""
+    ldflags = '-ldflags=-X github.com/algorand/indexer/version.Hash={}'.format(githash)
+    ldflags += ' -X github.com/algorand/indexer/version.Dirty={}'.format(dirty)
+    ldflags += ' -X github.com/algorand/indexer/version.CompileTime={}'.format(now)
+    ldflags += ' -X github.com/algorand/indexer/version.GitDecorateBase64={}'.format(base64.b64encode(desc.encode()).decode())
+    if release_version:
+        ldflags += ' -X github.com/algorand/indexer/version.ReleaseVersion={}'.format(release_version)
+    logger.debug('Hash=%r Dirty=%r CompileTime=%r decorate=%r ReleaseVersion=%r', githash, dirty, now, desc, release_version)
+    logger.debug('%s', ldflags)
+    return ldflags
+
+def compile(goos=None, goarch=None, ldflags=None):
     env = dict(os.environ)
     env['CGO_ENABLED'] = '0'
     if goos is not None:
         env['GOOS'] = goos
     if goarch is not None:
         env['GOARCH'] = goarch
-    subprocess.run(['go', 'build'], cwd='cmd/algorand-indexer', env=env).check_returncode()
+    cmd = ['go', 'build']
```

Comment:
is it possible to use `make cmd/algorand-indexer/algorand-indexer` instead of duplicating the flags setup?

Reply:
Maybe? I usually think of `make` as 'do the build if needed' and I want this to be more imperative 'do the build'; a preexisting build (and there will be 4 builds in a row for the os/arch combos) shouldn't prevent any of the building.

Reply:
OR, flip the script. Makefile calls this build script with options to only build for current os/arch